### PR TITLE
Upon running out of space in InvokeResponseMessage for additional responses, allocate another

### PR DESCRIPTION
**Pull request by tehampson**
This PR tries adding an `InvokeResponse` to a command into the `InvokeResponseMessage`. Upon failure due to lack of remaining space, it finalize the current `InvokeResponseMessage` and allocates another `InvokeResponseMessage`. It then retires to add that `InvokeResponse` into the new `InvokeResponseMessage`.

Change was also made to `TryAddResponseData`, instead of a template for `CommandData`, we made added Function template. This should reduce compiled code size.

Test:
* Using chip-repl script confirmed we are able to get to InvokeResponse messages
    ```
    dev_ctrl = devCtrl
    size = 700
    endpoint = 1
    command = Clusters.UnitTesting.Commands.TestBatchHelperRequest(50,size, ord('b'))
    invoke_request_1 = Clusters.Command.InvokeRequestInfo(endpoint, command)
    command = Clusters.UnitTesting.Commands.TestSecondBatchHelperRequest(0,size, ord('a'))
    invoke_request_2 = Clusters.Command.InvokeRequestInfo(endpoint, command)
    result = await dev_ctrl.SendBatchCommands(0x12344321, [invoke_request_1, invoke_request_2])
    ```

### CODE REVIEW

File: src/app/CommandHandler.h [hunk 1]
```diff
@@ -319,6 +299,48 @@ class CommandHandler
...
+    /**
+     * @brief Best effort to add InvokeResponse to InvokeResponseMessage.
+     * 
+     * Tries to add response using lambda. Upon failure to add response, attempts
+     * to rollback the InvokeResponseMessage to a known good state. If failure is due
+     * to insufficient space in the current InvokeResponseMessage:
+     *  - Finalizes the current InvokeResponseMessage.
+     *  - Allocates a new InvokeResponseMessage.
+     *  - Reattempts to add the InvokeResponse to the new InvokeResponseMessage.
+     *    the new InvokeResponseMessage.
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:311]:
```suggestion
```

File: src/app/CommandHandler.h [hunk 2]
```diff
@@ -319,6 +299,48 @@ class CommandHandler
...
+    CHIP_ERROR TryAddingResponse(Function && addResponseFunction)
+    {
+        mRollbackBackupValid = false;
+        CHIP_ERROR err       = addResponseFunction();
+        if (err == CHIP_NO_ERROR)
+        {
+            return CHIP_NO_ERROR;
+        }
+        ReturnErrorOnFailure(RollbackResponse());
+        // If we failed to an additional command due to lack of space in the
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:326]:
```suggestion
        // If we failed to add a command due to lack of space in the
```

File: src/app/CommandHandler.h [hunk 3]
```diff
@@ -319,6 +299,48 @@ class CommandHandler
...
+        {
+            return err;
+        }
+        ReturnErrorOnFailure(FinalizeInvokeResponseMessageAndReadyNext());
+        err                  = addResponseFunction();
+        mRollbackBackupValid = false;
+        if (err != CHIP_NO_ERROR)
+        {
+            // Over here we don't care about the return
+            RollbackResponse();
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:339]: This doesn't make sense to me.  We set mRollbackBackupValid to false above, so this will be a no-op, no?

I think these manual mRollbackBackupValid manipulations are pretty odd and at the very least need to be clearly documented.

File: src/app/CommandHandler.h [hunk 4]
```diff
@@ -563,6 +610,11 @@ class CommandHandler
     bool mGroupRequest                     = false;
     bool mBufferAllocated                  = false;
     bool mReserveSpaceForMoreChunkMessages = false;
+    // TODO(#30453): We should introduce breaking change where calls to add CommandData
+    // needs to use AddResponse, and not CommandHandler primatives directly using
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:614]:
```suggestion
    // need to use AddResponse, and not CommandHandler primatives directly using
```

File: src/app/CommandResponseSender.h [hunk 5]
```diff
@@ -123,6 +123,26 @@ class CommandResponseSender : public Messaging::ExchangeDelegate
         mChunks.AddToEnd(std::move(aPacket));
     }
 
+    /**
+     * @brief Sets final non-path specific StatusResponse error after sending all InvokeResponseMessages
```
Comment (bzbarsky-apple) [inline src/app/CommandResponseSender.h:127]: Is this after _sending_ all the InvokeResponseMessages?  Or after _queueing_ them?  This is not making sense to me as documented.

File: src/app/CommandResponseSender.h [hunk 6]
```diff
@@ -123,6 +123,26 @@ class CommandResponseSender : public Messaging::ExchangeDelegate
...
+     *
+     * Intended to be use for sending a single, non-path specific, error after we send all
```
Comment (bzbarsky-apple) [inline src/app/CommandResponseSender.h:129]:
```suggestion
     * Intended to be used for sending a single, non-path specific, error after we send all
```

File: src/app/CommandHandler.h (hunk 3 above)
Comment (tehampson) [reply src/app/CommandHandler.h:338]: It should be. I will try adding unit test that make sure it would have caught this issue.
```
        mRollbackBackupValid = false;
        err                  = addResponseFunction();
```

Comment (tehampson) [reply src/app/CommandHandler.h:340]: Unfortunately an attempt to encapsulate this and move it to the stack didn't work out :/
Working on adding documentation now

File: src/app/CommandHandler.h [hunk 7]
```diff
@@ -563,6 +616,11 @@ class CommandHandler
     bool mGroupRequest                     = false;
     bool mBufferAllocated                  = false;
     bool mReserveSpaceForMoreChunkMessages = false;
+    // TODO(#30453): We should introduce breaking change where calls to add CommandData
+    // need to use AddResponse, and not CommandHandler primatives directly using
+    // GetCommandDataIBTLVWriter.
+    bool mInternalCallToAddResponseData = false;
```
Comment (andy31415) [inline src/app/CommandHandler.h:622]: Should this be ScopedChangeOnly?

File: examples/lighting-app/silabs/openthread.gn [hunk 8]
```diff
@@ -24,6 +24,7 @@ default_args = {
   target_cpu = "arm"
   target_os = "freertos"
   chip_openthread_ftd = true
+  optimize_debug_level = "s"
```
Comment (tcarmelveilleux) [inline examples/lighting-app/silabs/openthread.gn:27]: Why this change?

File: src/app/CommandHandler.cpp [hunk 9]
```diff
@@ -655,12 +661,22 @@ CHIP_ERROR CommandHandler::PrepareInvokeResponseCommand(const CommandPathRegistr
...
-    mBackupState = mState;
+    if (!mInternalCallToAddResponseData && mState == State::AddedCommand)
+    {
+        // An attempt is being made to add CommandData InvokeResponse using primitive
+        // CommandHandler APIs. While not recommended, as this potentially leaves the
+        // CommandHandler in an incorrect state upon failure, this approach is permitted
+        // for legacy reasons. To maximize the likelihood of success, particularly when
+        // handling large amounts of data, we try to obtain a new, completely empty
+        // InvokeResponseMessage, as the existing one already has space occupied.
+        ReturnErrorOnFailure(FinalizeInvokeResponseMessageAndReadyNext());
```
Comment (tcarmelveilleux) [inline src/app/CommandHandler.cpp:672]: "XxxxxxAndPrepareNext()" ?

File: src/app/CommandHandler.cpp [hunk 10]
```diff
@@ -746,11 +766,31 @@ CHIP_ERROR CommandHandler::FinishStatus()
...
+{
+    if (mState != State::NewResponseMessage && mState != State::AddedCommand)
+    {
+        return;
+    }
+    if (mInvokeResponseBuilder.GetInvokeResponses().GetError() != CHIP_NO_ERROR ||
+        mInvokeResponseBuilder.GetError() != CHIP_NO_ERROR)
+    {
+        return;
+    }
```
Comment (tcarmelveilleux) [inline src/app/CommandHandler.cpp:779]: These 3 conditions can be converted to 3 separate `VerifyOrReturn`. The second if() with || could be split into 2 separate VerifyOrReturn.

File: src/app/CommandHandler.cpp [hunk 11]
```diff
@@ -746,11 +766,31 @@ CHIP_ERROR CommandHandler::FinishStatus()
...
+    mBackupState         = mState;
+    mRollbackBackupValid = true;
+}
+
 CHIP_ERROR CommandHandler::RollbackResponse()
 {
+    VerifyOrReturnError(mRollbackBackupValid, CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnError(mState == State::Preparing || mState == State::AddingCommand, CHIP_ERROR_INCORRECT_STATE);
+    mInvokeResponseBuilder.GetInvokeResponses().ResetError();
+    mInvokeResponseBuilder.ResetError();
```
Comment (tcarmelveilleux) [inline src/app/CommandHandler.cpp:790]: Would a generate `ResetErrors()` method helps avoid these two?

File: examples/lighting-app/silabs/openthread.gn (hunk 8 above)
Comment (jmartinez-silabs) [reply examples/lighting-app/silabs/openthread.gn:27]: I suggested this change to Terence because the lighting-app on MG12 was failing CI/running out of flash with this PR

Comment (tehampson) [reply examples/lighting-app/silabs/openthread.gn:27]: CI is failing due to space. Junior from Silabs mentioned over slack "(61a/ MG12 board) are starting to be tight in space", and "If it busts with your small changes it will bust for others soon"

I am the straw breaking the camels back. While I would like this to go in as a separate PR, because of code freeze I am trying to just get something in

File: src/app/CommandHandler.h [hunk 12]
```diff
@@ -319,6 +299,54 @@ class CommandHandler
...
+     * @brief Best effort to add InvokeResponse to InvokeResponseMessage.
+     *
+     * Tries to add response using lambda. Upon failure to add response, attempts
+     * to rollback the InvokeResponseMessage to a known good state. If failure is due
+     * to insufficient space in the current InvokeResponseMessage:
+     *  - Finalizes the current InvokeResponseMessage.
+     *  - Allocates a new InvokeResponseMessage.
+     *  - Reattempts to add the InvokeResponse to the new InvokeResponseMessage.
+     *
+     * addResponseFunction: A lambda function responsible for adding the response
```
Comment (tcarmelveilleux) [inline src/app/CommandHandler.h:312]: Please use `@param[in]` syntax.

File: src/app/CommandHandler.h [hunk 13]
```diff
@@ -563,6 +616,11 @@ class CommandHandler
     bool mGroupRequest                     = false;
     bool mBufferAllocated                  = false;
     bool mReserveSpaceForMoreChunkMessages = false;
+    // TODO(#30453): We should introduce breaking change where calls to add CommandData
+    // need to use AddResponse, and not CommandHandler primatives directly using
```
Comment (tcarmelveilleux) [inline src/app/CommandHandler.h:620]:
```suggestion
    // need to use AddResponse, and not CommandHandler primitives directly using
```

File: src/app/tests/TestCommandInteraction.cpp [hunk 14]
```diff
@@ -91,6 +92,30 @@ namespace app {
...
 
+struct ForcedSizeBuffer
+{
+    chip::Platform::ScopedMemoryBufferWithSize<uint8_t> mBuffer;
+
+    ForcedSizeBuffer(uint32_t size)
+    {
+        if (mBuffer.Alloc(size))
+        {
+            memset(mBuffer.Get(), 0x12, size);
```
Comment (tcarmelveilleux) [inline src/app/tests/TestCommandInteraction.cpp:103]: Why 0x12?

File: src/app/tests/TestCommandInteraction.cpp [hunk 15]
```diff
@@ -540,6 +573,27 @@ void TestCommandInteraction::AddInvokeResponseData(nlTestSuite * apSuite, void *
...
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    err = apCommandHandler->AllocateBuffer();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    uint32_t remainingSize = apCommandHandler->mInvokeResponseBuilder.GetWriter()->GetRemainingFreeLength();
+
+    // This number was derived by executing this method with aSizeToLeaveInBuffer = 100 and
+    // subsequently analyzing the remaining size to determine the overhead associated with calling
+    // AddResponseData with `ForcedSizeBuffer`.
+    uint32_t sizeNeededForAddingResponse = 27;
```
Comment (tcarmelveilleux) [inline src/app/tests/TestCommandInteraction.cpp:589]: This won't scale to changes to IM messages. Is there a different way to get the same test going without requiring exactly the right edge bounds? Like, could a bigger estimate be used, and then make the message slightly bigger than max, rather "exactly max", and slightly lower than max, and test both cases do the right thing? Otherwise, is it possible to reuse some primitives/utilities to build a response in a buffer, in such a way that the final buffer's .size() can be used to make this scale to changes to encoding?

This can be done in a follow-up.

File: src/app/tests/TestCommandInteraction.cpp [hunk 16]
```diff
@@ -1691,6 +1745,85 @@ void TestCommandInteraction::TestCommandHandlerAcceptMultipleCommands(nlTestSuit
     exchange->Close();
 }
 
+void TestCommandInteraction::TestCommandHandlerFillUpInvokeResponseMessageWhereSecondResponseIsStatusResponse(nlTestSuite * apSuite,
```
Comment (tcarmelveilleux) [inline src/app/tests/TestCommandInteraction.cpp:1748]: Suggest `_` between TestCommandHandler and `FillUpXxxxx` for readability, in all the enormously long method names.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/CommandHandler.h (hunk 12 above)
Comment (tehampson) [reply src/app/CommandHandler.h:312]: https://github.com/project-chip/connectedhomeip/actions/runs/7576265410/job/20634674678

```
/Users/runner/work/connectedhomeip/connectedhomeip/src/app/CommandHandler.h:312:15: error: parameter 'addResponseFunction:' not found in the function declaration [-Werror,-Wdocumentation]
     * @param addResponseFunction: A lambda function responsible for adding the
              ^~~~~~~~~~~~~~~~~~~~
/Users/runner/work/connectedhomeip/connectedhomeip/src/app/CommandHandler.h:312:15: note: did you mean 'addResponseFunction'?
     * @param addResponseFunction: A lambda function responsible for adding the
              ^~~~~~~~~~~~~~~~~~~~
              addResponseFunction
1 error generated.

Comment (tehampson) [reply src/app/CommandHandler.h:312]: I think I see what is wrong now. I will address this and see if it fixes the issue

File: src/app/tests/TestCommandInteraction.cpp (hunk 14 above)
Comment (tehampson) [reply src/app/tests/TestCommandInteraction.cpp:103]: Just some value to use. Added comment to mention that it doesn't have any significance

File: src/app/tests/TestCommandInteraction.cpp (hunk 15 above)
Comment (tehampson) [reply src/app/tests/TestCommandInteraction.cpp:589]: To test the extreme where we fill the buffer to the edge (to really test rollback). Yes this needs to exact max, this actually found a real bug that I fixed in this PR.

I have an idea of how IM layers can provide the values, but like you said in a follow up

File: src/app/CommandHandler.cpp [hunk 17]
```diff
@@ -801,6 +841,21 @@ CommandHandler::Handle::Handle(CommandHandler * handle)
     }
 }
 
+CHIP_ERROR CommandHandler::FinalizeInvokeResponseMessageAndReadyNext()
+{
+    ReturnErrorOnFailure(FinalizeInvokeResponseMessage(/* aHasMoreChunks = */ true));
+    CHIP_ERROR err = AllocateBuffer();
+    if (err == CHIP_NO_ERROR)
+    {
+        VerifyOrDie(mState == State::NewResponseMessage);
```
Comment (andy31415) [inline src/app/CommandHandler.cpp:850]: why is this being checked only if we do not have an allocation error? is it ok to be in a different state as long as allocation fails? can AllocateBuffer change state?

File: examples/lighting-app/silabs/openthread.gn (hunk 8 above)
Comment (tehampson) [reply examples/lighting-app/silabs/openthread.gn:27]: https://github.com/project-chip/connectedhomeip/pull/31555

Is in a standalond PR, plus now have a new idea to save some space

File: src/app/CommandHandler.h [hunk 18]
```diff
@@ -319,6 +299,54 @@ class CommandHandler
...
+     *  - Reattempts to add the InvokeResponse to the new InvokeResponseMessage.
+     *
+     * addResponseFunction: A lambda function responsible for adding the response
+     *    to the current InvokeResponseMessage.
+     */
+    template <typename Function>
+    CHIP_ERROR TryAddingResponse(Function && addResponseFunction)
+    {
+        // Invalidate any existing rollback backups. The addResponseFunction is
+        // expected to create a new backup during either PrepareInvokeResponseCommand
```
[the remark this answers is not in the document]
Comment (tehampson) [reply src/app/CommandHandler.h:319]: If we didn't give access to the primitives that allowed someone to muck around with the state of CommandHandler I would agree, but because we have provided access to that I think a Verify or die is riskier

File: src/app/CommandHandler.cpp (hunk 11 above)
Comment (tehampson) [reply src/app/CommandHandler.cpp:784]: If it was done more than once I would consider it, but this is the only spot where we are calling `ResetError`, so it would take more code to create a generic `ResetErrors`

File: src/app/CommandHandler.h [hunk 19]
```diff
@@ -332,15 +360,15 @@ class CommandHandler
...
-        CHIP_ERROR err = TryAddResponseData(aRequestCommandPath, aData);
-        if (err != CHIP_NO_ERROR)
-        {
-            // The state guarantees that either we can rollback or we don't have to rollback the buffer, so we don't care about the
-            // return value of RollbackResponse.
-            RollbackResponse();
-        }
-        return err;
+        // We are extracting the things needed by the CommandData template and keeping
+        // this function as minimal as possible to compiled code size.
```
Comment (tehampson) [inline src/app/CommandHandler.h:364]:
```suggestion
        // Extracting essential elements for CommandData templating to minimize compiled code size within this function.
```

File: src/app/CommandHandler.h [hunk 20]
```diff
@@ -549,24 +563,26 @@ class CommandHandler
...
-        ConcreteCommandPath responsePath = { aRequestCommandPath.mEndpointId, aRequestCommandPath.mClusterId,
-                                             CommandData::GetCommandId() };
-        ScopedChange internalCallToAddResponse(mInternalCallToAddResponseData, true);
-        ReturnErrorOnFailure(PrepareInvokeResponseCommand(responsePath, prepareParams));
+        ScopedChange<bool> internalCallToAddResponse(mInternalCallToAddResponseData, true);
+        ReturnErrorOnFailure(PrepareInvokeResponseCommand(aResponseCommandPath, prepareParams));
         TLV::TLVWriter * writer = GetCommandDataIBTLVWriter();
         VerifyOrReturnError(writer != nullptr, CHIP_ERROR_INCORRECT_STATE);
-        ReturnErrorOnFailure(DataModel::Encode(*writer, TLV::ContextTag(CommandDataIB::Tag::kFields), aData));
+        ReturnErrorOnFailure(encodeCommandDataFunction(*writer));
```
Comment (bzbarsky-apple) [inline src/app/CommandHandler.h:585]: So this change is not great; it's forcing all callers to handle the tags themselves, etc...  Is this actually reducing compiled code size?   It's not clear to me why, but I suggest:

1. Documenting in the caller _why_ factoring this way reduces code size.
2. Filing a followup to see whether we can have things better factored without regressing code size.

File: src/app/CommandResponseSender.h [hunk 21]
```diff
@@ -124,11 +124,14 @@ class CommandResponseSender : public Messaging::ExchangeDelegate
     }
 
     /**
-     * @brief Sets final non-path specific StatusResponse error after sending all InvokeResponseMessages
+     * @brief Queues a single, non-path-specific, error to be sent after all queued InvokeRequestMessages.
```
Comment (bzbarsky-apple) [inline src/app/CommandResponseSender.h:127]: So here's what I am not following:

1. Where do InvokeRequestMessages come here at all?  We're not queueing those here.
2. Why would we want to send any other messages at all if we know we have a non-path-specific (i.e. "this whole thing failed") error to send?

This logic is just not making sense to me.

Comment (tehampson) [reply src/app/CommandResponseSender.h:127]: 1. Queued InvokeResponseMessages (I recognize the typo, sorry) comes from `AddInvokeResponseToSend`

2. We have the InvokeResponseMessages, why not send them back? The application might want to know what happened to some of the commands before it failed.

File: src/app/CommandHandler.cpp [hunk 22]
```diff
@@ -781,6 +780,8 @@ CHIP_ERROR CommandHandler::RollbackResponse()
 {
     VerifyOrReturnError(mRollbackBackupValid, CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnError(mState == State::Preparing || mState == State::AddingCommand, CHIP_ERROR_INCORRECT_STATE);
+    mInvokeResponseBuilder.GetInvokeResponses().ResetError();
+    mInvokeResponseBuilder.ResetError();
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/CommandHandler.cpp]: Does this mean we need to have the invariant of:
  - `Checkpoint` may only be done on a non-error state (is this checked and enforced?)
  - This guarantees that any rollback rolls back into a non-error state
  
To me this sounds reasonable and would prevent more error-prone attempts to reset errors, however I would like to make sure we document and validate these invariants.

File: src/app/CommandHandler.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/CommandHandler.h:319]: The comment though says "it is expected to create a new backup". Is there a use case where we would not? Even if this is VerifyOrLog, at least to have something that checks assumptions. I believe even if we provide people with ability to customize things we should validate invariants.

or documentation should be worded in such a way that shows that "creating a new checkpoint" is actually optional and must be handled. If no checkpoint is made it looks like we should just fail RollbackResponse() so things will handle it already.

File: src/app/tests/TestCommandInteraction.cpp (hunk 14 above)
Comment (andy31415) [reply src/app/tests/TestCommandInteraction.cpp:104]: Fair random number generated by rolling a D20 .... :)

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/CommandHandler.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (tehampson) [reply src/app/CommandHandler.h:319]: > The comment though says "it is expected to create a new backup". Is there a use case where we would not?

Example of how this could fail today:
1. Some primitive comes along and leaves things in an incorrect state. Because we are in a bad state, we fail to create the backup.
2. We fail to allocate a buffer, so there is nothing to back up

[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/CommandHandler.h:319]: Resolving: this will (correctly) return an error regardless of what the addResponseFunction does. The code is better (i.e. can rollback and fix things) if we create a backup and the comment reflects that, however this does not need to be a strongly enforced check.
